### PR TITLE
Jsonnet vendoring

### PR DESCRIPTION
**Pull request by rollandf**
As suggested by @brancz , move the vendoring to each environment.

Is this breaking downstream project?

### CODE REVIEW

**Comment by metalmatze** [comment]
We don't really need a vendor folder per environment. They build on each others based on the same dependencies. 
How about simply having `vendor/` and `vendor-jsonnet/` or something here? :thinking:

**Comment by brancz** [comment]
Base environment should not have a lock file nor should it have the vendor checked in, only the operator environment should. The dev environment should have a local directory dependency pointing to the base environment.

**Comment by rollandf** [comment]
> We don't really need a vendor folder per environment. They build on each others based on the same dependencies.
> How about simply having `vendor/` and `vendor-jsonnet/` or something here? thinking

@metalmatze I have tried this way, but Go mode vendor will scan also  `vendor-jsonnet/`
You can try and see:
jb install --jsonnetpkg-home=jsonnet-vendor
go mod vendor

**Comment by brancz** [comment]
Maybe we can put the go and operator code in a subdirectory?

**Comment by rollandf** [comment]
> Maybe we can put the go and operator code in a subdirectory?

OK, what do you think the structure will be? Where the `jsonnetfile.json` should be?
Don't Go modules implies to have the go.mod in the root directory?

**Comment by brancz** [comment]
> Don't Go modules implies to have the go.mod in the root directory?

Is this really the case? If so I wasn't aware of it.

> Where the jsonnetfile.json should be?

I'm thinking also a separate directory that we just call `jsonnet`.

**Comment by rollandf** [comment]
> A module is a collection of Go packages stored in a file tree with a go.mod file at its root. 
from https://blog.golang.org/using-go-modules

**Comment by brancz** [comment]
The way I read this, `github.com/observatorium/configuration/operator` would be the "root" here of the mod, so `github.com/observatorium/configuration/operator/go.mod` should work just fine.

**Comment by rollandf** [comment]
OK, I will try this way

**Comment by rollandf** [comment]
@brancz PTAL
Do you want the jsonnetfile.json/jsonnetfile.lock.json under operator/jsonnet to be link to the one in root?
Still need to handle some changes in Makefile and circleci.

File: operator/jsonnet/jsonnetfile.json [hunk 1]
```diff
@@ -0,0 +1,89 @@
+{
```
Comment (brancz) [inline operator/jsonnet/jsonnetfile.json:1]: this jsonnetfile should only need to point at the base of this repo as a directory type package, like here: https://github.com/coreos/kube-prometheus/blob/master/jsonnetfile.json

the required dependencies are then transitively resolved

Comment (rollandf) [reply operator/jsonnet/jsonnetfile.json:1]: Here the structure is different, do I tried:

`{
  "version": 1,
  "dependencies": [
    {
      "source": {
        "local": {
          "directory": "../.."
        }
      },
      "version": ""
    }
  ],
  "legacyImports": true
}`

then jb install in `configuration/operator`, then it kind of messed up the directory.
It changed the `jsonnet` dir to a link:
`jsonnet -> /home/frolland/git/go/src/observatorium/configuration`

Comment (brancz) [reply operator/jsonnet/jsonnetfile.json:1]: That sounds like a bug in jb

Comment (brancz) [reply operator/jsonnet/jsonnetfile.json:1]: make sure to give the package a name still

Comment (rollandf) [reply operator/jsonnet/jsonnetfile.json:1]: Where should I give a name?

Comment (brancz) [reply operator/jsonnet/jsonnetfile.json:1]: `name` is a sibling of `version`

Comment (rollandf) [reply operator/jsonnet/jsonnetfile.json:1]: Added a name

Comment (brancz) [reply operator/jsonnet/jsonnetfile.json:1]: sorry, it's a sibling next to the dependency version, I should have been more clear:

```
{
  "version": 1,
  "dependencies": [
    {
      "source": {
        "local": {
          "directory": "../.."
        }
      },
      "version": "",
       "name": "configuration",
    }
  ],
  "legacyImports": true
}
```

Comment (rollandf) [reply operator/jsonnet/jsonnetfile.json:1]: I still have the issue with the link. If I have the jsonnetfile.json as above, under operator/jsonnet and then I run jb install, the dir turns to a link to the root, and the jsonnet dir content disappear, along with the jsonnetfile.json

Comment (rollandf) [reply operator/jsonnet/jsonnetfile.json:1]: Solved with jb built from master

**Comment by brancz** [comment]
This still has `WIP` is this ready for review?

**Comment by rollandf** [comment]
I still need to fix the Makefile.
Comments are welcome. :-)

**Comment by rollandf** [comment]
Ready for review.

Do I need to remove the `jsonnetfile.lock.json` in the root directory and add it to `.gitignore`?

**Comment by brancz** [comment]
> Do I need to remove the jsonnetfile.lock.json in the root directory and add it to .gitignore?

yes, we should only have the one from the operator environment as that's what produces the checked in vendoring

**Comment by rollandf** [comment]
@brancz  PTAL

**Comment by rollandf** [comment]
@brancz thanks for the guidance!

**Pull request merged**